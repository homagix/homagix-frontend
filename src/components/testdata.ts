import { Dish, Ingredient } from "@/types"

export const avocado = { id: "42", name: "Avocado", unit: "Stk" } as Ingredient
export const tomato = { id: "43", name: "Tomate", unit: "Stk" } as Ingredient

export const guacamole: Dish = {
  id: "5",
  name: "Guacamole",
  items: [
    { id: avocado.id, amount: 2 },
    { id: tomato.id, amount: 2 },
  ],
  alwaysOnList: false,
  images: ["guacamole.jpg"],
  isFavorite: false,
  isEditable: true,
}

export const tomatoSoup: Dish = {
  id: "7",
  name: "Tomatensuppe",
  items: [{ id: tomato.id, amount: 4 }],
  alwaysOnList: false,
  images: ["tomatoSoup.jpg"],
  isFavorite: false,
  isEditable: false,
}
