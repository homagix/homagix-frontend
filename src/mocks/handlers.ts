import { rest } from "msw"

export const handlers = [
  rest.get("http://localhost:8200/api/dishes", (req, res, ctx) => {
    return res(ctx.json({ dishes: [{ id: 1, name: "Pizza" }]}))
  }),
]
