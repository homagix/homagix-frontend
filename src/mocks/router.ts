import { defineComponent } from "vue"
import { createWebHistory } from "vue-router"
import { createRouter } from "vue-router"

export function createRouterMock(component: ReturnType<typeof defineComponent>, extraPaths: string[] = []) {
  const history = createWebHistory()
  const routes = ["/", ...extraPaths].map(path => ({ path, component }))
  return createRouter({ history, routes })
}
