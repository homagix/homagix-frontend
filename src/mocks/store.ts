import { vi } from "vitest"
import { createStore } from "vuex"

export function createStoreMock(state: object = {}, options: { actions?: string[]; mutations?: string[] } = {}) {
  function createMocksRecord(names?: string[]) {
    if (!names) {
      return undefined
    }
    return Object.assign({}, ...names.map(name => ({ [name]: vi.fn() })))
  }

  const actionMocks = createMocksRecord(options.actions)
  const mutationMocks = createMocksRecord(options.mutations)
  const store = createStore({
    state() {
      return state
    },
    actions: actionMocks,
    mutations: mutationMocks,
  })
  return { store, actionMocks, mutationMocks }
}
