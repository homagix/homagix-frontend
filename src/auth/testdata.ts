import { vi } from "vitest"
import { createStore } from "vuex"
import { mount } from "@vue/test-utils"
import { createRouterMock } from "../mocks/router"
import type { User } from "."

export const token =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjMiLCJmaXJzdE5hbWUiOiJhYmMiLCJpYXQiOjE2NDY5Nzg0NTUsImV4cCI6MTY0NzA2NDg1NX0.kClNCHxRnm_rvNpkdwdHgDsVebJfOzZ66Z_5epxkKNU"

export const user: User = { id: "123", name: "abc", accessCode: "456" }

export function mountComponent<C>(component: C) {
  const router = createRouterMock(component, ["/my", "/register"])
  const registerUserAction = vi.fn()
  const store = createStore({
    state() {
      return {}
    },
    actions: {
      REGISTER_USER: registerUserAction,
    },
  })
  const wrapper = mount(component, { global: { plugins: [store, router] } })
  return { wrapper, registerUserAction, router }
}
