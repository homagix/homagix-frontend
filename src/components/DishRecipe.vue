<script setup lang="ts">
import { computed } from "vue"
import { useStore } from "@/store"
import { useRoute, useRouter } from "vue-router"
import { Dish } from "@/types"
import { getImageUrl } from "@/api"
import AppButton from "./AppButton.vue"
import FavoriteButton from "./FavoriteButton.vue"

const store = useStore()
const route = useRoute()
const router = useRouter()

const dish = computed(() => store.state.dishes.find(dish => dish.id === route.params.id))

function ingredients() {
  return dish.value?.items.map(item => {
    return {
      ...store.state.ingredients?.find(ingredient => ingredient.id === item.id),
      amount: item.amount,
    }
  })
}

function backToList() {
  router.push({ name: "recipes" })
}

function image(dish: Dish) {
  return getImageUrl(dish.image)
}
</script>

<template>
  <section v-if="dish">
    <div class="title is-4">
      <span>{{ dish.name }}</span>
      <FavoriteButton :dish="dish" />
    </div>

    <div class="image-ingredients">
      <div class="image-wrapper">
        <o-icon icon="image"></o-icon>
        <img v-if="dish.image" :src="image(dish)" />
      </div>
      <div class="wrapper">
        <div id="ingredient-list">
          <template v-for="ingredient in ingredients()" :key="ingredient.id">
            <span>{{ ingredient.amount }}</span>
            <span>{{ ingredient.unit }}</span>
            <span>{{ ingredient.name }}</span>
          </template>
        </div>
      </div>
    </div>

    <div id="description">{{ dish.recipe || "Noch gibt es keine Beschreibung zu diesem Gericht" }}</div>

    <AppButton @click="backToList">Zur√ºck</AppButton>
  </section>
</template>

<style scoped lang="scss">
.title,
.image-ingredients {
  display: flex;
  justify-content: space-between;
}

.image-ingredients {
  position: relative;
  width: 100%;
  margin-bottom: 0.5rem;
  column-gap: 5px;

  .image-wrapper {
    width: 100%;
    position: relative;

    .icon {
      background: #f5f5f5;
      position: absolute;
      display: flex;
      width: 100%;
      height: 100%;
      font-size: 100px;
      color: #dddddd;
      z-index: 0;
    }

    img {
      position: relative;
    }
  }
}

#ingredient-list {
  display: inline-grid;
  grid-template-rows: auto;
  grid-template-columns: auto auto 1fr;
  border: 1px solid #888;
  padding: 3px;
  align-content: flex-start;

  > * {
    padding: 3px;
  }

  > *:nth-child(3n-2) {
    text-align: right;
  }
}

#description {
  margin-bottom: 1rem;
}
</style>
