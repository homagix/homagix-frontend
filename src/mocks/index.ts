import { mount, shallowMount } from "@vue/test-utils"
import { createRouterMock } from "./router"
import { createStoreMock, StoreMockOptions } from "./store"
import Oruga from "@oruga-ui/oruga-next"

type MountOptions = {
  shallow?: boolean
  props?: Record<string, any>
  state?: object
  storeOptions?: StoreMockOptions
  extraPaths?: string[]
}

export function mountComponent<C>(component: C, options: MountOptions) {
  const { store, actionMocks, mutationMocks } = createStoreMock(options.state || {}, options.storeOptions || {})
  const router = createRouterMock(component, options.extraPaths || [])
  const m = options.shallow === undefined || options.shallow ? shallowMount : mount
  const wrapper = m(component, { props: options.props || {}, global: { plugins: [store, router, Oruga] } })

  return { wrapper, store, router, actionMocks, mutationMocks }
}
