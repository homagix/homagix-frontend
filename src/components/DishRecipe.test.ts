import { describe, expect, it, vi } from "vitest"
import DishRecipe from "./DishRecipe.vue"
import { mountComponent } from "@/mocks"
import { avocado, guacamole, tomato, tomatoSoup } from "./testdata"

const options = (id: string) => ({
  props: { id },
  state: { dishes: [guacamole, tomatoSoup], ingredients: [avocado, tomato] },
  extraPaths: ["/recipes"],
})

describe("DishRecipe", () => {
  it("should render as expected", () => {
    const { wrapper } = mountComponent(DishRecipe, options(guacamole.id))
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("shows a button which returns back to the list", async () => {
    const { wrapper, router } = mountComponent(DishRecipe, options(guacamole.id))
    router.push = vi.fn()
    const button = wrapper.find("*[icon='list']")
    expect(button.exists()).toBe(true)
    await button.trigger("click")
    expect(router.push).toHaveBeenCalledWith("/recipes")
  })

  it("has an edit button if the current user is authorized", () => {
    const { wrapper } = mountComponent(DishRecipe, options(guacamole.id))
    expect(wrapper.find("*[icon='edit']").exists()).toBe(true)
  })

  it("doesn't have edit, save or cancel buttons if the user is not authorized", () => {
    const { wrapper } = mountComponent(DishRecipe, options(tomatoSoup.id))
    expect(wrapper.find("*[icon='edit']").exists()).toBe(false)
    expect(wrapper.find("*[icon='save']").exists()).toBe(false)
    expect(wrapper.find("*[icon='xmark']").exists()).toBe(false)
  })

  it("doesn't have save or cancel buttons when not in edit mode", () => {
    const { wrapper } = mountComponent(DishRecipe, options(guacamole.id))
    expect(wrapper.find("*[icon='save']").exists()).toBe(false)
    expect(wrapper.find("*[icon='xmark']").exists()).toBe(false)
  })

  it("shows a save and a cancel button in edit mode", async () => {
    const { wrapper } = mountComponent(DishRecipe, options(guacamole.id))
    await wrapper.find("*[icon='edit']").trigger("click")
    expect(wrapper.find("*[icon='save']").exists()).toBe(true)
    expect(wrapper.find("*[icon='xmark']").exists()).toBe(true)
  })
})
