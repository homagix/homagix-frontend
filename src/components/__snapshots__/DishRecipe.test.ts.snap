// Vitest Snapshot v1

exports[`DishRecipe > should render as expected 1`] = `
"<section data-v-6da79d3c=\\"\\">
  <div class=\\"title is-4\\" data-v-6da79d3c=\\"\\"><span data-v-6da79d3c=\\"\\">Guacamole</span>
    <favorite-button-stub dish=\\"[object Object]\\" data-v-6da79d3c=\\"\\"></favorite-button-stub>
  </div>
  <div class=\\"image-ingredients\\" data-v-6da79d3c=\\"\\">
    <div class=\\"image-wrapper\\" data-v-6da79d3c=\\"\\">
      <o-icon-stub icon=\\"image\\" clickable=\\"false\\" spin=\\"false\\" both=\\"false\\" data-v-6da79d3c=\\"\\"></o-icon-stub>
      <!--v-if-->
    </div>
    <div class=\\"ingredients-list-container\\" data-v-6da79d3c=\\"\\">
      <p class=\\"title is-6\\" data-v-6da79d3c=\\"\\">Zutaten:</p>
      <ingredients-list-stub ingredients=\\"[object Object],[object Object]\\" data-v-6da79d3c=\\"\\"></ingredients-list-stub>
    </div>
  </div>
  <vue-showdown-stub markdown=\\"Noch gibt es keine Beschreibung zu diesem Gericht\\" tag=\\"div\\" flavor=\\"github\\" options=\\"[object Object]\\" vuetemplate=\\"false\\" vuetemplatedata=\\"[object Object]\\" class=\\"description\\" data-v-6da79d3c=\\"\\"></vue-showdown-stub>
  <div class=\\"buttons\\" data-v-6da79d3c=\\"\\">
    <app-button-stub icon=\\"list\\" data-v-6da79d3c=\\"\\"></app-button-stub>
    <!--v-if-->
    <app-button-stub icon=\\"edit\\" data-v-6da79d3c=\\"\\"></app-button-stub>
    <!--v-if-->
  </div>
</section>"
`;
